### PR TITLE
Giving transforms by argument to shader

### PR DESCRIPTION
**Pull request by davidpiuva**
Reducing side effects in code to make it easy to know which arguments are actually used.

### CODE REVIEW

**Comment by marcusnaslund** [comment]
#1682 merged

File: source/draw/Map.ooc [hunk 1]
```diff
@@ -53,22 +53,13 @@ _Map: class {
 }
 
 Map: abstract class {
-	model: FloatTransform3D { get set }
-	view: FloatTransform3D { get set }
-	projection: FloatTransform3D { get set }
-	textureTransform: FloatTransform3D { get set }
```
Comment (davidpiuva) [inline source/draw/Map.ooc]: These variables are global side effects between all modules using the same shaders.

File: source/draw/Map.ooc [hunk 2]
```diff
@@ -53,22 +53,13 @@ _Map: class {
...
-		this projection = FloatTransform3D identity
-		this textureTransform = FloatTransform3D identity
-	}
+	init: func
 	free: override func {
 		this _bindings free()
 		super()
 	}
-	use: virtual func (forbiddenInput: Pointer)
+	use: abstract func (forbiddenInput: Pointer, positionTransform, textureTransform: FloatTransform3D)
```
Comment (davidpiuva) [inline source/draw/Map.ooc]: The transforms are now given by argument.

File: source/draw/gpu/opengl/OpenGLCanvas.ooc [hunk 3]
```diff
@@ -54,7 +54,7 @@ OpenGLCanvas: class extends OpenGLSurface {
 					gpuMap add("texture0", image)
 			}
 		}
-		gpuMap use(this _target)
+		gpuMap use(this _target, projection * view * model, textureTransform)
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLCanvas.ooc]: Doing the multiplication here makes it easier to refactor model-view-projection.

File: source/draw/gpu/opengl/OpenGLContext.ooc [hunk 4]
```diff
@@ -80,17 +80,15 @@ OpenGLContext: class extends GpuContext {
 	drawQuad: func { this _renderer drawQuad() }
 	drawLines: func (pointList: VectorList<FloatPoint2D>, projection: FloatTransform3D, pen: Pen) {
 		positions := pointList pointer as Float*
-		this _linesShader projection = projection
 		this _linesShader add("color", pen color normalized)
-		this _linesShader use(null)
+		this _linesShader use(null, projection, FloatTransform3D identity)
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Now these calls don't even have to perform any matrix multiplications on the CPU.

File: source/draw/gpu/opengl/OpenGLMap.ooc [hunk 5]
```diff
@@ -104,11 +104,10 @@ OpenGLMap: class extends Map {
 
 OpenGLMapTransform: class extends OpenGLMap {
 	init: func (fragmentSource: String, context: OpenGLContext) { super(This vertexSource, fragmentSource, context) }
-	use: override func (forbiddenInput: Pointer) {
-		finalTransform := this projection * this view * this model
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLMap.ooc]: The multiplications are removed from the inside.

File: source/ui/x11/UnixWindow.ooc [hunk 6]
```diff
@@ -74,11 +74,11 @@ UnixWindow: class extends UnixWindowBase {
...
 		sourceSize := image size toFloatVector2D()
-		map textureTransform = GpuCanvas _createTextureTransform(image size, IntBox2D new(image size))
-		map model = FloatTransform3D createScaling(sourceSize x / 2.0f, sourceSize y / 2.0f, 0.0f)
-		map view = this _openGLWindow _createView(sourceSize, FloatTransform3D identity)
-		map projection = this _openGLWindow _createProjection(sourceSize, 0.0f)
-		map use(null)
+		textureTransform := GpuCanvas _createTextureTransform(image size, IntBox2D new(image size))
+		model := FloatTransform3D createScaling(sourceSize x / 2.0f, sourceSize y / 2.0f, 0.0f)
+		view := this _openGLWindow _createView(sourceSize, FloatTransform3D identity)
+		projection := this _openGLWindow _createProjection(sourceSize, 0.0f)
```
Comment (davidpiuva) [inline source/ui/x11/UnixWindow.ooc]: Transforms are stored in temporary variables and then multiplied. Anything that is not used can be removed from the transform product.

**Comment by davidpiuva** [comment]
@emilwestergren Peer review?

**Comment by davidpiuva** [comment]
@marcusnaslund All checks passed.

**Pull request merged**